src/main/kotlin/sudoklin/solver/SudokuSolver.kt

```diff
@@ -183,4 +184,181 @@ class SudokuSolver(initialSudoku: Sudoku, val log: Boolean = false) {
+                        possiblePositions.add(rowIndex)
+                    }
+                }
+            }
+        }
+
+        if (log) File("./$timestamp.log").appendText("   $possiblePositions \r\n")
+
+        return possiblePositions
+    }
+
+    fun getPossiblePositionsForNumberInGroup(number: Int, groupIndex: Int): List<Pair<Int, Int>> {
+        val group = sudoku.groups[groupIndex]
+        val possiblePositions: MutableList<Pair<Int, Int>> = mutableListOf<Pair<Int, Int>>()
+        val alreadyInGoup = sudoku.groups[groupIndex]!!.containsSolvedNumber(number)
+
+        if (log) File("./$timestamp.log").appendText("$number in group $groupIndex: " + group.toString() + "   $alreadyInGoup")
+
+        if (!alreadyInGoup) {
+            for (rowIndex in 0..8) {
+                for (columnIndex in 0..8) {
+                    if (sudoku.getCell(rowIndex, columnIndex).groupIndex == groupIndex) {
+                        if (!sudoku.getCell(rowIndex, columnIndex).isSolved()) {
+                            val alreadyInRow = sudoku.rows[rowIndex]!!.containsSolvedNumber(number)
+                            val alreadyInColumn = sudoku.columns[columnIndex]!!.containsSolvedNumber(number)
+
+                            if (! (alreadyInRow || alreadyInColumn)) {
+                                possiblePositions.add(Pair(rowIndex, columnIndex))
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (log) File("./$timestamp.log").appendText("   $possiblePositions \r\n")
+
+        return possiblePositions
+    }
+
+    fun solve(): NewShinySudoku {
+        var iteration = 0
+        if (log) File("./$timestamp.log").appendText("INITIAL" + sudoku.toString())
+        while(! sudoku.isSolved()) {
+            // TODO refactor:
+            // use simple method until no more progress,
+            // then use next more complicated method,
+            // anytime a number is found, start over from step 1 again
+
+//            /*1*/ fillOnlyPossibleValueForCells()
+//            if (log) File("./$timestamp.log").appendText("Step $iteration a:" + workingStack.last().puzzle.toString())
+
+            /*2*/ fillNumbersToTheirOnlyPossibleCell()
+            if (log) File("./$timestamp.log").appendText("Step $iteration b:" + sudoku.toString())
+//            /*3*/ findPairsInCandidatesAndEliminateOthers()
+//            if (log) File("./$timestamp.log").appendText("Step $iteration c:" + workingStack.last().puzzle.toString())
+            iteration++
+        }
+        return sudoku
+    }
+
+    private fun fillNumbersToTheirOnlyPossibleCell() {
+        for (number in 1..9) {
+            for (i in 0..8) {
+                val possiblePositionsInRow = getPossiblePositionsForNumberInRow(number, i)
+                if (possiblePositionsInRow.size == 1) {
+                    sudoku.rows[i]!!.cells[possiblePositionsInRow[0]].candidates.clear()
```

Comment:
we are always doing `clear` and then `add`. Let's have one convenience function which wraps this.